mountpoint-s3/src/fs.rs

```diff
@@ -55,7 +55,91 @@ enum FileHandleType<Client: ObjectClient, Runtime> {
         request: AsyncMutex<Option<PrefetchGetObject<Client, Runtime>>>,
         etag: ETag,
     },
-    Write(AsyncMutex<Option<UploadRequest<Client, WriteHandle>>>),
+    Write {
+        request: AsyncMutex<UploadState<Client>>,
+        handle: WriteHandle,
+    },
+}
+
+#[derive(Debug)]
+enum UploadState<Client: ObjectClient> {
+    InProgress(UploadRequest<Client>),
+    Completed,
+    Failed,
+}
+
+impl<Client: ObjectClient> UploadState<Client> {
+    async fn write(&mut self, offset: i64, data: &[u8], key: &str) -> Result<u32, libc::c_int> {
+        let upload = match self {
+            Self::InProgress(upload) => upload,
+            Self::Completed => {
+                warn!(key, "upload already completed");
+                return Err(libc::EIO);
+            }
+            Self::Failed => {
+                warn!(key, "upload already aborted");
+                return Err(libc::EIO);
+            }
+        };
+        match upload.write(offset, data).await {
+            Ok(len) => Ok(len as u32),
+            Err(e) => {
+                error!("write failed: {e}");
+                match e {
+                    UploadWriteError::PutRequestFailed(_) => {
+                        // Abort the request.
+                        *self = Self::Failed;
+                    }
+                    UploadWriteError::OutOfOrderWrite { .. } => {}
+                };
+                Err(e.into())
+            }
+        }
+    }
+
+    async fn complete(&mut self, key: &str) -> Result<(), libc::c_int> {
+        let upload = match std::mem::replace(self, Self::Completed) {
```

Comment:
This is kinda weird because it can replace `Self::Failed` with `Self::Completed`. I think that only messes up error messages in some edge cases, but probably we should try to get it right anyway.